distutils is not available on some non-Linux OS's

tgolly opened the pull request:
##### Issue Type:  Bugfix Pull Request
##### Ansible Version:  tested on v1.8.2.  can reproduce on devel if required.
##### Environment:  Solaris
##### Summary:

The "service" module is broken on Solaris since 30d6713.

distutils.LooseService() is only referenced in the LinuxService class, this ensures it's only loaded on Linux platforms.

After this patch the service module works again on Solaris.
##### Steps To Reproduce:

Use the service module on a Solaris machine:

```
service: name=/network/ntp state=restarted
```
##### Expected Results:

ntp daemon is restarted
##### Actual Results:

```
ImportError: No module named distutils.version
```

tgolly committed:
distutils is not available on some non-Linux OS's

abadger commented:
Where are you getting your python version from?  Which version of python is it? Is distutils present but not distutils.version?  ( python -c 'import distutils' works but python -c 'import distutils.version' does not)  distutils is a part of python's stdlib so it should be there unless the python version on your OS is being crippled by your vendor.  Without distutils, installing third party python packages from pypi wouldn't work as well as several means of querying for information about the system. Trying to google for whether solaris is lacking distuitls in general turns up only one relevant page from stackoverflow so this doesn't seem quite right.... perhaps there's a separate Solaris python-libs or python-dev or something package that needs to be installed to put distutils on the system?

tgolly commented:
Hi abadger, you are right!  It appears distutils is available in the devel package.  

```
# grep distutils/__init__.py SUNWPython-devel/pkgmap
1 f none lib/python2.6/distutils/__init__.py 0644 root bin 670 57754 1340772982
1 f none lib/python2.6/distutils/__init__.pyc 0644 root bin 437 30473 1340772985
```

Unfortunately this has a chilling effect on our ability to use Ansible, we can't install devel packages on the production servers.

I'd really like my one-line patch to be considered, as distutils is only required for the Linux platform.  It means we wouldn't need to have SUNWPython-devel installed on our prod servers which keeps the security folks happy.

abadger commented:
<nod>   Yeah, I see your dilemma but I also don't want to put us in a position where we're avoiding making use of stdlib modules because some vendors have crippled their packages unnecessarily.  (For what it's worth, older Ubuntu Linux also split distutils into a separate, non-default devel package and users had no choice but to install that package until Ubuntu moved distutils into the main package). 

How about this -- if you rework your patch so that it avoids the import on Solaris instead of only importing on Linux and add a comment about the split packages on Solaris, I'll merge it.  But I can't guarantee that we won't find we need to use distutils unconditionally somewhere in the future.

Sound fair?

tgolly committed:
Load distutils on all platforms EXCEPT Solaris.

Solaris doesn't ship distutils with with the default Python package.  This patch fixes "service" on Solaris since 30d6713.

tgolly commented:
Yep, sounds fair and I agree.  Patch updated.  Thanks!

abadger commented:
Thanks for the fix!

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
